[CI] SearchTimeoutIT testAggsTimeout failing

Failed 15 times in the past 3 days

**Build scan:**
https://gradle-enterprise.elastic.co/s/7hpvqannlo7wu/tests/:server:internalClusterTest/org.elasticsearch.search.SearchTimeoutIT/testAggsTimeout

**Reproduction line:**
```
gradlew ':server:internalClusterTest' --tests "org.elasticsearch.search.SearchTimeoutIT.testAggsTimeout" -Dtests.seed=E8376C653A8A9230 -Dtests.locale=sk-SK -Dtests.timezone=Asia/Dili -Druntime.java=20
```

**Applicable branches:**
main

**Reproduces locally?:**
No

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.search.SearchTimeoutIT&tests.test=testAggsTimeout

**Failure excerpt:**
```
java.lang.AssertionError: 
Expected: a value greater than <0L>
     but: <0L> was equal to <0L>

  at __randomizedtesting.SeedInfo.seed([E8376C653A8A9230:4E5B9455DBFC30CE]:0)
  at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
  at org.junit.Assert.assertThat(Assert.java:956)
  at org.junit.Assert.assertThat(Assert.java:923)
  at org.elasticsearch.search.SearchTimeoutIT.testAggsTimeout(SearchTimeoutIT.java:83)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
  at java.lang.reflect.Method.invoke(Method.java:578)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1623)

```

--- a/server/src/internalClusterTest/java/org/elasticsearch/search/SearchTimeoutIT.java
+++ b/server/src/internalClusterTest/java/org/elasticsearch/search/SearchTimeoutIT.java
@@ -9,6 +9,8 @@
 
 package org.elasticsearch.search;
 
+import com.carrotsearch.randomizedtesting.annotations.Repeat;
+
 import org.apache.lucene.util.ThreadInterruptedException;
 import org.elasticsearch.ElasticsearchException;
 import org.elasticsearch.action.search.SearchRequestBuilder;
@@ -54,7 +56,7 @@ public class SearchTimeoutIT extends ESIntegTestCase {
     @Override
     protected void setupSuiteScopeCluster() throws Exception {
         super.setupSuiteScopeCluster();
-        indexRandom(true, "test", randomIntBetween(5, 30));
+        indexRandom(true, "test", randomIntBetween(20, 50));
     }
 
     @Override
@@ -63,6 +65,7 @@ public class SearchTimeoutIT extends ESIntegTestCase {
         scriptExecutions.set(0);
     }
 
+    @Repeat(iterations = 100)
     public void testTopHitsTimeout() {
         SearchRequestBuilder searchRequestBuilder = prepareSearch("test").setTimeout(new TimeValue(100, TimeUnit.MILLISECONDS))
             .setQuery(scriptQuery(new Script(ScriptType.INLINE, "mockscript", SCRIPT_NAME, Collections.emptyMap())));
@@ -77,8 +80,8 @@ public class SearchTimeoutIT extends ESIntegTestCase {
         });
     }
 
+    @Repeat(iterations = 100)
     public void testAggsTimeout() {
-        indexRandom(true, "test", randomIntBetween(5, 30));
         SearchRequestBuilder searchRequestBuilder = prepareSearch("test").setTimeout(new TimeValue(100, TimeUnit.MILLISECONDS))
             .setSize(0)
             .setQuery(scriptQuery(new Script(ScriptType.INLINE, "mockscript", SCRIPT_NAME, Collections.emptyMap())))
@@ -120,7 +123,7 @@ public class SearchTimeoutIT extends ESIntegTestCase {
                 // This ensures that one document is always returned before the timeout happens.
                 // Also, don't sleep any further to avoid slowing down the test excessively.
                 // A timeout on a specific slice of a single shard is enough.
-                if (scriptExecutions.incrementAndGet() == 1) {
+                if (scriptExecutions.getAndIncrement() == 1) {
                     try {
                         Thread.sleep(500);
                     } catch (InterruptedException e) {

--- a/server/src/internalClusterTest/java/org/elasticsearch/search/SearchTimeoutIT.java
+++ b/server/src/internalClusterTest/java/org/elasticsearch/search/SearchTimeoutIT.java
@@ -9,8 +9,9 @@
 
 package org.elasticsearch.search;
 
+import org.apache.lucene.util.ThreadInterruptedException;
 import org.elasticsearch.ElasticsearchException;
-import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.action.search.SearchRequestBuilder;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.core.TimeValue;
 import org.elasticsearch.plugins.Plugin;
@@ -20,21 +21,25 @@ import org.elasticsearch.script.ScriptType;
 import org.elasticsearch.search.aggregations.bucket.terms.StringTerms;
 import org.elasticsearch.search.aggregations.bucket.terms.TermsAggregationBuilder;
 import org.elasticsearch.test.ESIntegTestCase;
+import org.elasticsearch.test.hamcrest.ElasticsearchAssertions;
 
 import java.util.Collection;
 import java.util.Collections;
 import java.util.Map;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
 import java.util.function.Function;
 
-import static org.elasticsearch.action.support.WriteRequest.RefreshPolicy.IMMEDIATE;
 import static org.elasticsearch.index.query.QueryBuilders.scriptQuery;
 import static org.elasticsearch.search.SearchTimeoutIT.ScriptedTimeoutPlugin.SCRIPT_NAME;
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.greaterThan;
 
 @ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.SUITE)
+@ESIntegTestCase.SuiteScopeTestCase
 public class SearchTimeoutIT extends ESIntegTestCase {
+
+    private static final AtomicInteger scriptExecutions = new AtomicInteger(0);
 
     @Override
     protected Collection<Class<? extends Plugin>> nodePlugins() {
@@ -46,53 +51,55 @@ public class SearchTimeoutIT extends ESIntegTestCase {
         return Settings.builder().put(super.nodeSettings(nodeOrdinal, otherSettings)).build();
     }
 
-    private void indexDocs() {
-        for (int i = 0; i < 32; i++) {
-            prepareIndex("test").setId(Integer.toString(i)).setSource("field", "value").get();
-        }
-        refresh("test");
+    @Override
+    protected void setupSuiteScopeCluster() throws Exception {
+        super.setupSuiteScopeCluster();
+        indexRandom(true, "test", randomIntBetween(5, 30));
     }
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/98369")
+    @Override
+    public void tearDown() throws Exception {
+        super.tearDown();
+        scriptExecutions.set(0);
+    }
+
     public void testTopHitsTimeout() {
-        indexDocs();
-        SearchResponse searchResponse = prepareSearch("test").setTimeout(new TimeValue(10, TimeUnit.MILLISECONDS))
-            .setQuery(scriptQuery(new Script(ScriptType.INLINE, "mockscript", SCRIPT_NAME, Collections.emptyMap())))
-            .get();
-        assertThat(searchResponse.isTimedOut(), equalTo(true));
-        assertEquals(0, searchResponse.getShardFailures().length);
-        assertEquals(0, searchResponse.getFailedShards());
-        assertThat(searchResponse.getSuccessfulShards(), greaterThan(0));
-        assertEquals(searchResponse.getSuccessfulShards(), searchResponse.getTotalShards());
-        assertThat(searchResponse.getHits().getTotalHits().value(), greaterThan(0L));
-        assertThat(searchResponse.getHits().getHits().length, greaterThan(0));
+        SearchRequestBuilder searchRequestBuilder = prepareSearch("test").setTimeout(new TimeValue(100, TimeUnit.MILLISECONDS))
+            .setQuery(scriptQuery(new Script(ScriptType.INLINE, "mockscript", SCRIPT_NAME, Collections.emptyMap())));
+        ElasticsearchAssertions.assertResponse(searchRequestBuilder, searchResponse -> {
+            assertThat(searchResponse.isTimedOut(), equalTo(true));
+            assertEquals(0, searchResponse.getShardFailures().length);
+            assertEquals(0, searchResponse.getFailedShards());
+            assertThat(searchResponse.getSuccessfulShards(), greaterThan(0));
+            assertEquals(searchResponse.getSuccessfulShards(), searchResponse.getTotalShards());
+            assertThat(searchResponse.getHits().getTotalHits().value(), greaterThan(0L));
+            assertThat(searchResponse.getHits().getHits().length, greaterThan(0));
+        });
     }
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/98053")
     public void testAggsTimeout() {
-        indexDocs();
-        SearchResponse searchResponse = prepareSearch("test").setTimeout(new TimeValue(10, TimeUnit.MILLISECONDS))
+        indexRandom(true, "test", randomIntBetween(5, 30));
+        SearchRequestBuilder searchRequestBuilder = prepareSearch("test").setTimeout(new TimeValue(100, TimeUnit.MILLISECONDS))
             .setSize(0)
             .setQuery(scriptQuery(new Script(ScriptType.INLINE, "mockscript", SCRIPT_NAME, Collections.emptyMap())))
-            .addAggregation(new TermsAggregationBuilder("terms").field("field.keyword"))
-            .get();
-        assertThat(searchResponse.isTimedOut(), equalTo(true));
-        assertEquals(0, searchResponse.getShardFailures().length);
-        assertEquals(0, searchResponse.getFailedShards());
-        assertThat(searchResponse.getSuccessfulShards(), greaterThan(0));
-        assertEquals(searchResponse.getSuccessfulShards(), searchResponse.getTotalShards());
-        assertThat(searchResponse.getHits().getTotalHits().value(), greaterThan(0L));
-        assertEquals(searchResponse.getHits().getHits().length, 0);
-        StringTerms terms = searchResponse.getAggregations().get("terms");
-        assertEquals(1, terms.getBuckets().size());
-        StringTerms.Bucket bucket = terms.getBuckets().get(0);
-        assertEquals("value", bucket.getKeyAsString());
-        assertThat(bucket.getDocCount(), greaterThan(0L));
+            .addAggregation(new TermsAggregationBuilder("terms").field("field.keyword"));
+        ElasticsearchAssertions.assertResponse(searchRequestBuilder, searchResponse -> {
+            assertThat(searchResponse.isTimedOut(), equalTo(true));
+            assertEquals(0, searchResponse.getShardFailures().length);
+            assertEquals(0, searchResponse.getFailedShards());
+            assertThat(searchResponse.getSuccessfulShards(), greaterThan(0));
+            assertEquals(searchResponse.getSuccessfulShards(), searchResponse.getTotalShards());
+            assertThat(searchResponse.getHits().getTotalHits().value(), greaterThan(0L));
+            assertEquals(0, searchResponse.getHits().getHits().length);
+            StringTerms terms = searchResponse.getAggregations().get("terms");
+            assertEquals(1, terms.getBuckets().size());
+            StringTerms.Bucket bucket = terms.getBuckets().get(0);
+            assertEquals("value", bucket.getKeyAsString());
+            assertThat(bucket.getDocCount(), greaterThan(0L));
+        });
     }
 
-    public void testPartialResultsIntolerantTimeout() throws Exception {
-        prepareIndex("test").setId("1").setSource("field", "value").setRefreshPolicy(IMMEDIATE).get();
-
+    public void testPartialResultsIntolerantTimeout() {
         ElasticsearchException ex = expectThrows(
             ElasticsearchException.class,
             prepareSearch("test").setTimeout(new TimeValue(10, TimeUnit.MILLISECONDS))
@@ -100,6 +107,7 @@ public class SearchTimeoutIT extends ESIntegTestCase {
                 .setAllowPartialSearchResults(false) // this line causes timeouts to report failures
         );
         assertTrue(ex.toString().contains("Time exceeded"));
+        assertEquals(504, ex.status().getStatus());
     }
 
     public static class ScriptedTimeoutPlugin extends MockScriptPlugin {
@@ -108,10 +116,16 @@ public class SearchTimeoutIT extends ESIntegTestCase {
         @Override
         public Map<String, Function<Map<String, Object>, Object>> pluginScripts() {
             return Collections.singletonMap(SCRIPT_NAME, params -> {
-                try {
-                    Thread.sleep(500);
-                } catch (InterruptedException e) {
-                    throw new RuntimeException(e);
+                // sleep only once per test, but only after executing the script once without sleeping.
+                // This ensures that one document is always returned before the timeout happens.
+                // Also, don't sleep any further to avoid slowing down the test excessively.
+                // A timeout on a specific slice of a single shard is enough.
+                if (scriptExecutions.incrementAndGet() == 1) {
+                    try {
+                        Thread.sleep(500);
+                    } catch (InterruptedException e) {
+                        throw new ThreadInterruptedException(e);
+                    }
                 }
                 return true;
             });

--- a/server/src/internalClusterTest/java/org/elasticsearch/search/SearchTimeoutIT.java
+++ b/server/src/internalClusterTest/java/org/elasticsearch/search/SearchTimeoutIT.java
@@ -15,6 +15,7 @@ import org.apache.lucene.util.ThreadInterruptedException;
 import org.elasticsearch.ElasticsearchException;
 import org.elasticsearch.action.search.SearchRequestBuilder;
 import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.core.SuppressForbidden;
 import org.elasticsearch.core.TimeValue;
 import org.elasticsearch.plugins.Plugin;
 import org.elasticsearch.script.MockScriptPlugin;
@@ -65,6 +66,7 @@ public class SearchTimeoutIT extends ESIntegTestCase {
         scriptExecutions.set(0);
     }
 
+    @SuppressForbidden(reason = "just a test")
     @Repeat(iterations = 100)
     public void testTopHitsTimeout() {
         SearchRequestBuilder searchRequestBuilder = prepareSearch("test").setTimeout(new TimeValue(100, TimeUnit.MILLISECONDS))
@@ -80,6 +82,7 @@ public class SearchTimeoutIT extends ESIntegTestCase {
         });
     }
 
+    @SuppressForbidden(reason = "just a test")
     @Repeat(iterations = 100)
     public void testAggsTimeout() {
         SearchRequestBuilder searchRequestBuilder = prepareSearch("test").setTimeout(new TimeValue(100, TimeUnit.MILLISECONDS))

--- a/server/src/internalClusterTest/java/org/elasticsearch/search/SearchTimeoutIT.java
+++ b/server/src/internalClusterTest/java/org/elasticsearch/search/SearchTimeoutIT.java
@@ -181,10 +181,10 @@ public class SearchTimeoutIT extends ESIntegTestCase {
     }
 
     /**
-     * Query builder that produces a Lucene Query which throws a {@link org.elasticsearch.search.internal.ContextIndexSearcher.TimeExceededException}
-     * while scoring documents. This helps make this test not time dependent, otherwise it would be unpredictable when exactly the timeout
-     * happens: before even starting to score documents, or while scoring them? This is rather important if we want to test that we are
-     * able to return partial results on timeout.
+     * Query builder that produces a Lucene Query which throws a
+     * {@link org.elasticsearch.search.internal.ContextIndexSearcher.TimeExceededException} before or while scoring documents.
+     * This helps make this test not time dependent, otherwise it would be unpredictable when exactly the timeout happens, which is
+     * rather important if we want to test that we are able to return partial results on timeout.
      */
     public static final class BulkScorerTimeoutQuery extends AbstractQueryBuilder<BulkScorerTimeoutQuery> {
 

--- a/server/src/internalClusterTest/java/org/elasticsearch/search/SearchTimeoutIT.java
+++ b/server/src/internalClusterTest/java/org/elasticsearch/search/SearchTimeoutIT.java
@@ -9,32 +9,46 @@
 
 package org.elasticsearch.search;
 
-import com.carrotsearch.randomizedtesting.annotations.Repeat;
-
-import org.apache.lucene.util.ThreadInterruptedException;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.BulkScorer;
+import org.apache.lucene.search.ConstantScoreScorer;
+import org.apache.lucene.search.ConstantScoreWeight;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.LeafCollector;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.Scorable;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.ScorerSupplier;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.util.Bits;
 import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.TransportVersion;
 import org.elasticsearch.action.search.SearchRequestBuilder;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
 import org.elasticsearch.common.settings.Settings;
-import org.elasticsearch.core.SuppressForbidden;
 import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.index.query.AbstractQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.SearchExecutionContext;
 import org.elasticsearch.plugins.Plugin;
-import org.elasticsearch.script.MockScriptPlugin;
-import org.elasticsearch.script.Script;
-import org.elasticsearch.script.ScriptType;
+import org.elasticsearch.plugins.SearchPlugin;
 import org.elasticsearch.search.aggregations.bucket.terms.StringTerms;
 import org.elasticsearch.search.aggregations.bucket.terms.TermsAggregationBuilder;
+import org.elasticsearch.search.internal.ContextIndexSearcher;
 import org.elasticsearch.test.ESIntegTestCase;
 import org.elasticsearch.test.hamcrest.ElasticsearchAssertions;
+import org.elasticsearch.xcontent.XContentBuilder;
 
+import java.io.IOException;
 import java.util.Collection;
 import java.util.Collections;
-import java.util.Map;
+import java.util.List;
 import java.util.concurrent.TimeUnit;
-import java.util.concurrent.atomic.AtomicInteger;
-import java.util.function.Function;
 
-import static org.elasticsearch.index.query.QueryBuilders.scriptQuery;
-import static org.elasticsearch.search.SearchTimeoutIT.ScriptedTimeoutPlugin.SCRIPT_NAME;
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.greaterThan;
 
@@ -42,11 +56,9 @@ import static org.hamcrest.Matchers.greaterThan;
 @ESIntegTestCase.SuiteScopeTestCase
 public class SearchTimeoutIT extends ESIntegTestCase {
 
-    private static final AtomicInteger scriptExecutions = new AtomicInteger(0);
-
     @Override
     protected Collection<Class<? extends Plugin>> nodePlugins() {
-        return Collections.singleton(ScriptedTimeoutPlugin.class);
+        return Collections.singleton(BulkScorerTimeoutQueryPlugin.class);
     }
 
     @Override
@@ -60,17 +72,26 @@ public class SearchTimeoutIT extends ESIntegTestCase {
         indexRandom(true, "test", randomIntBetween(20, 50));
     }
 
-    @Override
-    public void tearDown() throws Exception {
-        super.tearDown();
-        scriptExecutions.set(0);
+    public void testTopHitsTimeoutBeforeCollecting() {
+        // setting the timeout is necessary only because we check that if a TimeExceededException is thrown, a timeout was set
+        SearchRequestBuilder searchRequestBuilder = prepareSearch("test").setTimeout(new TimeValue(10, TimeUnit.SECONDS))
+            .setQuery(new BulkScorerTimeoutQuery(false));
+        ElasticsearchAssertions.assertResponse(searchRequestBuilder, searchResponse -> {
+            assertThat(searchResponse.isTimedOut(), equalTo(true));
+            assertEquals(0, searchResponse.getShardFailures().length);
+            assertEquals(0, searchResponse.getFailedShards());
+            assertThat(searchResponse.getSuccessfulShards(), greaterThan(0));
+            assertEquals(searchResponse.getSuccessfulShards(), searchResponse.getTotalShards());
+            // timeout happened before we could collect any doc, total hits is 0 and no hits are returned
+            assertEquals(0, searchResponse.getHits().getTotalHits().value());
+            assertEquals(0, searchResponse.getHits().getHits().length);
+        });
     }
 
-    @SuppressForbidden(reason = "just a test")
-    @Repeat(iterations = 100)
-    public void testTopHitsTimeout() {
-        SearchRequestBuilder searchRequestBuilder = prepareSearch("test").setTimeout(new TimeValue(100, TimeUnit.MILLISECONDS))
-            .setQuery(scriptQuery(new Script(ScriptType.INLINE, "mockscript", SCRIPT_NAME, Collections.emptyMap())));
+    public void testTopHitsTimeoutWhileCollecting() {
+        // setting the timeout is necessary only because we check that if a TimeExceededException is thrown, a timeout was set
+        SearchRequestBuilder searchRequestBuilder = prepareSearch("test").setTimeout(new TimeValue(10, TimeUnit.SECONDS))
+            .setQuery(new BulkScorerTimeoutQuery(true));
         ElasticsearchAssertions.assertResponse(searchRequestBuilder, searchResponse -> {
             assertThat(searchResponse.isTimedOut(), equalTo(true));
             assertEquals(0, searchResponse.getShardFailures().length);
@@ -82,12 +103,31 @@ public class SearchTimeoutIT extends ESIntegTestCase {
         });
     }
 
-    @SuppressForbidden(reason = "just a test")
-    @Repeat(iterations = 100)
-    public void testAggsTimeout() {
-        SearchRequestBuilder searchRequestBuilder = prepareSearch("test").setTimeout(new TimeValue(100, TimeUnit.MILLISECONDS))
-            .setSize(0)
-            .setQuery(scriptQuery(new Script(ScriptType.INLINE, "mockscript", SCRIPT_NAME, Collections.emptyMap())))
+    public void testAggsTimeoutBeforeCollecting() {
+        SearchRequestBuilder searchRequestBuilder = prepareSearch("test").setSize(0)
+            // setting the timeout is necessary only because we check that if a TimeExceededException is thrown, a timeout was set
+            .setTimeout(new TimeValue(10, TimeUnit.SECONDS))
+            .setQuery(new BulkScorerTimeoutQuery(false))
+            .addAggregation(new TermsAggregationBuilder("terms").field("field.keyword"));
+        ElasticsearchAssertions.assertResponse(searchRequestBuilder, searchResponse -> {
+            assertThat(searchResponse.isTimedOut(), equalTo(true));
+            assertEquals(0, searchResponse.getShardFailures().length);
+            assertEquals(0, searchResponse.getFailedShards());
+            assertThat(searchResponse.getSuccessfulShards(), greaterThan(0));
+            assertEquals(searchResponse.getSuccessfulShards(), searchResponse.getTotalShards());
+            assertEquals(0, searchResponse.getHits().getTotalHits().value());
+            assertEquals(0, searchResponse.getHits().getHits().length);
+            StringTerms terms = searchResponse.getAggregations().get("terms");
+            // timeout happened before we could collect any doc, total hits is 0 and no buckets are returned
+            assertEquals(0, terms.getBuckets().size());
+        });
+    }
+
+    public void testAggsTimeoutWhileCollecting() {
+        SearchRequestBuilder searchRequestBuilder = prepareSearch("test").setSize(0)
+            // setting the timeout is necessary only because we check that if a TimeExceededException is thrown, a timeout was set
+            .setTimeout(new TimeValue(10, TimeUnit.SECONDS))
+            .setQuery(new BulkScorerTimeoutQuery(true))
             .addAggregation(new TermsAggregationBuilder("terms").field("field.keyword"));
         ElasticsearchAssertions.assertResponse(searchRequestBuilder, searchResponse -> {
             assertThat(searchResponse.isTimedOut(), equalTo(true));
@@ -105,36 +145,173 @@ public class SearchTimeoutIT extends ESIntegTestCase {
         });
     }
 
-    public void testPartialResultsIntolerantTimeout() {
+    public void testPartialResultsIntolerantTimeoutBeforeCollecting() {
         ElasticsearchException ex = expectThrows(
             ElasticsearchException.class,
-            prepareSearch("test").setTimeout(new TimeValue(10, TimeUnit.MILLISECONDS))
-                .setQuery(scriptQuery(new Script(ScriptType.INLINE, "mockscript", SCRIPT_NAME, Collections.emptyMap())))
+            prepareSearch("test")
+                // setting the timeout is necessary only because we check that if a TimeExceededException is thrown, a timeout was set
+                .setTimeout(new TimeValue(10, TimeUnit.SECONDS))
+                .setQuery(new BulkScorerTimeoutQuery(false))
                 .setAllowPartialSearchResults(false) // this line causes timeouts to report failures
         );
         assertTrue(ex.toString().contains("Time exceeded"));
-        assertEquals(504, ex.status().getStatus());
+        assertEquals(429, ex.status().getStatus());
     }
 
-    public static class ScriptedTimeoutPlugin extends MockScriptPlugin {
-        static final String SCRIPT_NAME = "search_timeout";
+    public void testPartialResultsIntolerantTimeoutWhileCollecting() {
+        ElasticsearchException ex = expectThrows(
+            ElasticsearchException.class,
+            prepareSearch("test")
+                // setting the timeout is necessary only because we check that if a TimeExceededException is thrown, a timeout was set
+                .setTimeout(new TimeValue(10, TimeUnit.SECONDS))
+                .setQuery(new BulkScorerTimeoutQuery(true))
+                .setAllowPartialSearchResults(false) // this line causes timeouts to report failures
+        );
+        assertTrue(ex.toString().contains("Time exceeded"));
+        assertEquals(429, ex.status().getStatus());
+    }
+
+    public static final class BulkScorerTimeoutQueryPlugin extends Plugin implements SearchPlugin {
+        @Override
+        public List<QuerySpec<?>> getQueries() {
+            return Collections.singletonList(new QuerySpec<QueryBuilder>("timeout", BulkScorerTimeoutQuery::new, parser -> {
+                throw new UnsupportedOperationException();
+            }));
+        }
+    }
+
+    /**
+     * Query builder that produces a Lucene Query which throws a {@link org.elasticsearch.search.internal.ContextIndexSearcher.TimeExceededException}
+     * while scoring documents. This helps make this test not time dependent, otherwise it would be unpredictable when exactly the timeout
+     * happens: before even starting to score documents, or while scoring them? This is rather important if we want to test that we are
+     * able to return partial results on timeout.
+     */
+    public static final class BulkScorerTimeoutQuery extends AbstractQueryBuilder<BulkScorerTimeoutQuery> {
+
+        private final boolean partialResults;
+
+        BulkScorerTimeoutQuery(boolean partialResults) {
+            this.partialResults = partialResults;
+        }
+
+        BulkScorerTimeoutQuery(StreamInput in) throws IOException {
+            super(in);
+            this.partialResults = in.readBoolean();
+        }
 
         @Override
-        public Map<String, Function<Map<String, Object>, Object>> pluginScripts() {
-            return Collections.singletonMap(SCRIPT_NAME, params -> {
-                // sleep only once per test, but only after executing the script once without sleeping.
-                // This ensures that one document is always returned before the timeout happens.
-                // Also, don't sleep any further to avoid slowing down the test excessively.
-                // A timeout on a specific slice of a single shard is enough.
-                if (scriptExecutions.getAndIncrement() == 1) {
-                    try {
-                        Thread.sleep(500);
-                    } catch (InterruptedException e) {
-                        throw new ThreadInterruptedException(e);
-                    }
+        protected void doWriteTo(StreamOutput out) throws IOException {
+            out.writeBoolean(partialResults);
+        }
+
+        @Override
+        protected void doXContent(XContentBuilder builder, Params params) {}
+
+        @Override
+        protected Query doToQuery(SearchExecutionContext context) {
+            return new Query() {
+                @Override
+                public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) {
+                    return new ConstantScoreWeight(this, boost) {
+                        @Override
+                        public boolean isCacheable(LeafReaderContext ctx) {
+                            return false;
+                        }
+
+                        @Override
+                        public ScorerSupplier scorerSupplier(LeafReaderContext context) {
+                            return new ScorerSupplier() {
+                                @Override
+                                public BulkScorer bulkScorer() {
+                                    if (partialResults == false) {
+                                        ((ContextIndexSearcher) searcher).throwTimeExceededException();
+                                    }
+                                    final int maxDoc = context.reader().maxDoc();
+                                    return new BulkScorer() {
+                                        @Override
+                                        public int score(LeafCollector collector, Bits acceptDocs, int min, int max) throws IOException {
+                                            max = Math.min(max, maxDoc);
+                                            collector.setScorer(new Scorable() {
+                                                @Override
+                                                public float score() {
+                                                    return 1f;
+                                                }
+                                            });
+                                            for (int doc = min; doc < max; ++doc) {
+                                                if (acceptDocs == null || acceptDocs.get(doc)) {
+                                                    collector.collect(doc);
+                                                    // collect one doc, then throw a timeout, this ensures partial results will be returned
+                                                    ((ContextIndexSearcher) searcher).throwTimeExceededException();
+                                                }
+                                            }
+                                            throw new AssertionError(
+                                                "Should have thrown a time exceeded exception: [min: " + min + " - max: " + max + "]"
+                                            );
+                                        }
+
+                                        @Override
+                                        public long cost() {
+                                            return maxDoc;
+                                        }
+                                    };
+                                }
+
+                                @Override
+                                public Scorer get(long leadCost) {
+                                    assert false;
+                                    return new ConstantScoreScorer(score(), scoreMode, DocIdSetIterator.all(context.reader().maxDoc()));
+                                }
+
+                                @Override
+                                public long cost() {
+                                    assert false;
+                                    return context.reader().maxDoc();
+                                }
+                            };
+                        }
+                    };
                 }
-                return true;
-            });
+
+                @Override
+                public String toString(String field) {
+                    return "timeout query";
+                }
+
+                @Override
+                public void visit(QueryVisitor visitor) {
+                    visitor.visitLeaf(this);
+                }
+
+                @Override
+                public boolean equals(Object obj) {
+                    return sameClassAs(obj);
+                }
+
+                @Override
+                public int hashCode() {
+                    return classHash();
+                }
+            };
+        }
+
+        @Override
+        protected boolean doEquals(BulkScorerTimeoutQuery other) {
+            return false;
+        }
+
+        @Override
+        protected int doHashCode() {
+            return 0;
+        }
+
+        @Override
+        public String getWriteableName() {
+            return "timeout";
+        }
+
+        @Override
+        public TransportVersion getMinimalSupportedVersion() {
+            return null;
         }
     }
 }

